File: scripts/validate_submission.py
#!/usr/bin/env python3
"""Validate Tama Builder against public skills-only submission constraints."""

from __future__ import annotations

import argparse
import json
import re
import sys
import xml.etree.ElementTree as ET
from pathlib import Path, PurePosixPath
from urllib.parse import urlparse


CATEGORIES = {
    "Productivity",
    "Creativity",
    "Developer Tools",
    "Business & Operations",
    "Data & Analytics",
    "Communication",
    "Education & Research",
    "Security",
    "Finance",
    "Healthcare",
    "Travel",
    "Entertainment",
    "Other",
}
IMAGE_SUFFIXES = {".png", ".jpg", ".jpeg", ".webp", ".svg"}
SEMVER = re.compile(
    r"^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)"
    r"(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?"
    r"(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$"
)


class SubmissionError(ValueError):
    """Raised when the package is not ready for directory submission."""


def require(condition: bool, message: str) -> None:
    if not condition:
        raise SubmissionError(message)


def read_json(path: Path) -> object:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise SubmissionError(f"Cannot read valid JSON from {path}: {error}") from error


def require_text(value: object, field: str, maximum: int, *, one_line: bool = True) -> str:
    require(
        isinstance(value, str) and value.strip() == value and bool(value),
        f"{field} must be non-empty text",
    )
    require(
        len(value) <= maximum,
        f"{field} must be {maximum} characters or fewer (found {len(value)})",
    )
    if one_line:
        require("\n" not in value and "\r" not in value, f"{field} must fit on one line")
    return value


def require_https_url(value: object, field: str) -> None:
    text = require_text(value, field, 1024)
    parsed = urlparse(text)
    require(
        parsed.scheme == "https" and bool(parsed.netloc),
        f"{field} must be a public HTTPS URL",
    )
    require(
        parsed.username is None and parsed.password is None,
        f"{field} must not embed credentials",
    )


def parse_svg_dimensions(path: Path) -> tuple[float, float]:
    try:
        root = ET.parse(path).getroot()
    except (OSError, ET.ParseError) as error:
        raise SubmissionError(f"Cannot parse SVG {path}: {error}") from error

    require(root.tag.rsplit("}", 1)[-1] == "svg", f"{path} must have an <svg> root")

    view_box = root.get("viewBox")
    if view_box:
        try:
            values = [float(value) for value in re.split(r"[\s,]+", view_box.strip())]
        except ValueError as error:
            raise SubmissionError(f"{path} has a non-numeric viewBox") from error
        require(len(values) == 4, f"{path} viewBox must contain four numbers")
        width, height = values[2], values[3]
    else:
        width_text = root.get("width", "")
        height_text = root.get("height", "")
        numeric = re.compile(r"^(?:\d+(?:\.\d*)?|\.\d+)$")
        require(
            bool(numeric.fullmatch(width_text)),
            f"{path} width must be numeric and unitless",
        )
        require(
            bool(numeric.fullmatch(height_text)),
            f"{path} height must be numeric and unitless",
        )
        width, height = float(width_text), float(height_text)

    require(width >= 48 and height >= 48, f"{path} must be at least 48x48")
    require(width == height, f"{path} must be square (found {width}x{height})")
    return width, height


def validate_asset(root: Path, value: object, field: str) -> None:
    text = require_text(value, field, 2048)
    require(text.startswith("./"), f"{field} must start with ./")
    relative = PurePosixPath(text[2:])
    require(
        not relative.is_absolute() and ".." not in relative.parts,
        f"{field} must stay inside the plugin",
    )
    path = root.joinpath(*relative.parts)
    require(
        path.is_file() and not path.is_symlink(),
        f"{field} must reference a regular packaged file",
    )
    require(
        path.suffix.lower() in IMAGE_SUFFIXES,
        f"{field} uses an unsupported image type",
    )
    require(
        path.stat().st_size <= 5 * 1024 * 1024,
        f"{field} must not exceed 5 MiB",
    )
    if path.suffix.lower() == ".svg":
        parse_svg_dimensions(path)


def validate_manifest(root: Path) -> tuple[dict[str, object], dict[str, object]]:
    manifest = read_json(root / ".codex-plugin" / "plugin.json")
    package = read_json(root / "package.json")
    require(isinstance(manifest, dict), "Plugin manifest must contain a JSON object")
    require(isinstance(package, dict), "package.json must contain a JSON object")

    name = require_text(manifest.get("name"), "name", 64)
    require(
        bool(re.fullmatch(r"[A-Za-z0-9][A-Za-z0-9_-]*", name)),
        "name has unsupported characters",
    )
    version = require_text(manifest.get("version"), "version", 64)
    require(bool(SEMVER.fullmatch(version)), "version must use semantic versioning")
    require(
        package.get("version") == version,
        "package.json and plugin manifest versions must match",
    )
    require(
        package.get("license") == "Apache-2.0",
        "package.json license must be Apache-2.0",
    )

    require(
        "mcpServers" not in manifest and not (root / ".mcp.json").exists(),
        "Skills-only ZIP must not include MCP configuration",
    )
    require(
        "apps" not in manifest and not (root / ".app.json").exists(),
        "Skills-only ZIP must not include app configuration",
    )

    interface = manifest.get("interface")
    require(isinstance(interface, dict), "interface must contain a JSON object")
    require_text(interface.get("displayName"), "interface.displayName", 30)
    require_text(interface.get("shortDescription"), "interface.shortDescription", 30)
    require_text(
        interface.get("longDescription"),
        "interface.longDescription",
        4000,
        one_line=False,
    )
    require_text(interface.get("developerName"), "interface.developerName", 80)
    require(interface.get("category") in CATEGORIES, "interface.category is unsupported")

    capabilities = interface.get("capabilities", [])
    require(
        isinstance(capabilities, list) and len(capabilities) <= 20,
        "interface.capabilities must contain at most 20 entries",
    )
    for index, capability in enumerate(capabilities):
        require_text(capability, f"interface.capabilities[{index}]", 120)

    prompts = interface.get("defaultPrompt", [])
    require(
        isinstance(prompts, list) and len(prompts) <= 3,
        "interface.defaultPrompt must contain at most three entries",
    )
    normalized_prompts: set[str] = set()
    for index, prompt in enumerate(prompts):
        prompt_text = require_text(prompt, f"interface.defaultPrompt[{index}]", 128)
        require(
            "@" not in prompt_text,
            f"interface.defaultPrompt[{index}] must not contain an app @mention",
        )
        normalized = " ".join(prompt_text.split()).casefold()
        require(
            normalized not in normalized_prompts,
            "interface.defaultPrompt entries must be unique",
        )
        normalized_prompts.add(normalized)

    for field in (
        "websiteURL",
        "privacyPolicyURL",
        "termsOfServiceURL",
        "supportURL",
    ):
        if field in interface:
            require_https_url(interface[field], f"interface.{field}")

    validate_asset(root, interface.get("logo"), "interface.logo")
    validate_asset(root, interface.get("composerIcon"), "interface.composerIcon")
    return manifest, package


def validate_case_text(case: dict[str, object], field: str, case_id: str) -> None:
    require_text(case.get(field), f"{case_id}.{field}", 4000, one_line=False)


def validate_evals(root: Path) -> tuple[int, int]:
    data = read_json(root / "evals" / "cases.json")
    require(isinstance(data, dict), "evals/cases.json must contain a JSON object")
    positive = data.get("positive_cases")
    negative = data.get("negative_cases")
    require(
        isinstance(positive, list) and len(positive) >= 5,
        "At least five positive test cases are required",
    )
    require(
        isinstance(negative, list) and len(negative) >= 3,
        "At least three negative test cases are required",
    )

    seen_ids: set[str] = set()
    for case in positive:
        require(isinstance(case, dict), "Each positive test case must be an object")
        case_id = require_text(case.get("id"), "positive case id", 120)
        require(case_id not in seen_ids, f"Duplicate test case id: {case_id}")
        seen_ids.add(case_id)
        validate_case_text(case, "prompt", case_id)
        require_text(case.get("skill"), f"{case_id}.skill", 120)
        behavior = case.get("expected_behavior")
        require(
            isinstance(behavior, list) and bool(behavior),
            f"{case_id}.expected_behavior must be a non-empty list",
        )
        for index, item in enumerate(behavior):
            require_text(
                item,
                f"{case_id}.expected_behavior[{index}]",
                1000,
                one_line=False,
            )
        validate_case_text(case, "expected_result_shape", case_id)
        validate_case_text(case, "fixture_data", case_id)

    for case in negative:
        require(isinstance(case, dict), "Each negative test case must be an object")
        case_id = require_text(case.get("id"), "negative case id", 120)
        require(case_id not in seen_ids, f"Duplicate test case id: {case_id}")
        seen_ids.add(case_id)
        validate_case_text(case, "prompt", case_id)
        validate_case_text(case, "expected_behavior", case_id)
        validate_case_text(case, "reason", case_id)

    portal = read_json(root / "submission" / "portal.json")
    require(isinstance(portal, dict), "submission/portal.json must contain a JSON object")
    require(portal.get("submissionType") == "Skills only", "submissionType must be Skills only")
    listing = portal.get("listing")
    require(isinstance(listing, dict), "submission listing must be an object")
    for field in (
        "websiteURL",
        "privacyPolicyURL",
        "termsOfServiceURL",
        "supportURL",
    ):
        require_https_url(listing.get(field), f"submission.listing.{field}")
    validate_case_text(portal, "releaseNotes", "submission")

    positive_ids = {case["id"] for case in positive}
    negative_ids = {case["id"] for case in negative}
    selected_positive = portal.get("positiveCaseIds")
    selected_negative = portal.get("negativeCaseIds")
    require(
        isinstance(selected_positive, list) and len(selected_positive) >= 5,
        "submission.positiveCaseIds must select at least five positive cases",
    )
    require(
        isinstance(selected_negative, list) and len(selected_negative) >= 3,
        "submission.negativeCaseIds must select at least three negative cases",
    )
    for index, case_id in enumerate(selected_positive):
        require_text(case_id, f"submission.positiveCaseIds[{index}]", 120)
    for index, case_id in enumerate(selected_negative):
        require_text(case_id, f"submission.negativeCaseIds[{index}]", 120)
    require(
        len(selected_positive) == len(set(selected_positive)),
        "submission.positiveCaseIds must be unique",
    )
    require(
        len(selected_negative) == len(set(selected_negative)),
        "submission.negativeCaseIds must be unique",
    )
    require(
        set(selected_positive).issubset(positive_ids),
        "submission.positiveCaseIds contains an unknown case",
    )
    require(
        set(selected_negative).issubset(negative_ids),
        "submission.negativeCaseIds contains an unknown case",
    )

    return len(positive), len(negative)


def validate_repository(root: Path) -> tuple[dict[str, object], dict[str, object]]:
    root = root.resolve()
    manifest, package = validate_manifest(root)
    positive_count, negative_count = validate_evals(root)
    print(
        f"Submission validation passed: {manifest['name']} {manifest['version']} "
        f"({positive_count} positive, {negative_count} negative cases)"
    )
    return manifest, package


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "root",
        nargs="?",
        type=Path,
        default=Path(__file__).resolve().parents[1],
    )
    args = parser.parse_args()
    try:
        validate_repository(args.root)
    except SubmissionError as error:
        print(f"Submission validation failed: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/validate-submission.mjs
#!/usr/bin/env node

import { existsSync, lstatSync, readdirSync, readFileSync } from "node:fs";
import { dirname, extname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import {
  placeholderNames,
  validatePublicExampleUrl,
  validateTemplateMatch,
} from "./lib/mcp-config.mjs";

const CATEGORIES = new Set([
  "Productivity",
  "Creativity",
  "Developer Tools",
  "Business & Operations",
  "Data & Analytics",
  "Communication",
  "Education & Research",
  "Security",
  "Finance",
  "Healthcare",
  "Travel",
  "Entertainment",
  "Other",
]);
const IMAGE_SUFFIXES = new Set([".png", ".jpg", ".jpeg", ".webp", ".svg"]);
const ROOT_DEFAULT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const SEMVER =
  /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/u;

export class SubmissionError extends Error {}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

export function requireCondition(condition, message) {
  if (!condition) {
    throw new SubmissionError(message);
  }
}

export function readJson(path) {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    throw new SubmissionError(`Cannot read valid JSON from ${path}: ${error.message}`);
  }
}

function requireText(value, field, maximum, { oneLine = true } = {}) {
  requireCondition(
    typeof value === "string" && value.trim() === value && value.length > 0,
    `${field} must be non-empty text`,
  );
  requireCondition(
    value.length <= maximum,
    `${field} must be ${maximum} characters or fewer (found ${value.length})`,
  );
  if (oneLine) {
    requireCondition(!/[\r\n]/u.test(value), `${field} must fit on one line`);
  }
  return value;
}

function requireHttpsUrl(value, field) {
  const text = requireText(value, field, 1024);
  let parsed;
  try {
    parsed = new URL(text);
  } catch {
    throw new SubmissionError(`${field} must be a public HTTPS URL`);
  }
  requireCondition(
    parsed.protocol === "https:" && Boolean(parsed.hostname),
    `${field} must be a public HTTPS URL`,
  );
  requireCondition(!parsed.username && !parsed.password, `${field} must not embed credentials`);
}

function attributeValue(attributes, name) {
  const escaped = name.replace(/[.*+?^${}()|[\]\\]/gu, "\\$&");
  const match = attributes.match(new RegExp(`(?:^|\\s)${escaped}\\s*=\\s*(["'])(.*?)\\1`, "su"));
  return match?.[2];
}

function parseSvgDimensions(path) {
  let text;
  try {
    text = readFileSync(path, "utf8");
  } catch (error) {
    throw new SubmissionError(`Cannot parse SVG ${path}: ${error.message}`);
  }
  const root = text.match(/<svg\b([^>]*)>/su);
  requireCondition(root !== null, `${path} must have an <svg> root`);

  const viewBox = attributeValue(root[1], "viewBox");
  let width;
  let height;
  if (viewBox) {
    const values = viewBox
      .trim()
      .split(/[\s,]+/u)
      .map(Number);
    requireCondition(
      values.length === 4 && values.every(Number.isFinite),
      `${path} viewBox must contain four numbers`,
    );
    width = values[2];
    height = values[3];
  } else {
    const widthText = attributeValue(root[1], "width") ?? "";
    const heightText = attributeValue(root[1], "height") ?? "";
    const numeric = /^(?:\d+(?:\.\d*)?|\.\d+)$/u;
    requireCondition(numeric.test(widthText), `${path} width must be numeric and unitless`);
    requireCondition(numeric.test(heightText), `${path} height must be numeric and unitless`);
    width = Number(widthText);
    height = Number(heightText);
  }

  requireCondition(width >= 48 && height >= 48, `${path} must be at least 48x48`);
  requireCondition(width === height, `${path} must be square (found ${width}x${height})`);
  return [width, height];
}

function validateAsset(root, value, field) {
  const text = requireText(value, field, 2048);
  requireCondition(text.startsWith("./"), `${field} must start with ./`);
  const parts = text.slice(2).split("/");
  requireCondition(
    !parts.includes("..") && !parts.includes("") && !text.slice(2).startsWith("/"),
    `${field} must stay inside the plugin`,
  );
  const path = join(root, ...parts);
  let metadata;
  try {
    metadata = lstatSync(path);
  } catch {
    throw new SubmissionError(`${field} must reference a regular packaged file`);
  }
  requireCondition(
    metadata.isFile() && !metadata.isSymbolicLink(),
    `${field} must reference a regular packaged file`,
  );
  const suffix = extname(path).toLowerCase();
  requireCondition(IMAGE_SUFFIXES.has(suffix), `${field} uses an unsupported image type`);
  requireCondition(metadata.size <= 5 * 1024 * 1024, `${field} must not exceed 5 MiB`);
  if (suffix === ".svg") {
    parseSvgDimensions(path);
  }
}

function validateManifest(root) {
  const manifest = readJson(join(root, ".codex-plugin", "plugin.json"));
  const packageJson = readJson(join(root, "package.json"));
  requireCondition(isObject(manifest), "Plugin manifest must contain a JSON object");
  requireCondition(isObject(packageJson), "package.json must contain a JSON object");

  const name = requireText(manifest.name, "name", 64);
  requireCondition(/^[A-Za-z0-9][A-Za-z0-9_-]*$/u.test(name), "name has unsupported characters");
  const version = requireText(manifest.version, "version", 64);
  requireCondition(SEMVER.test(version), "version must use semantic versioning");
  requireCondition(
    packageJson.version === version,
    "package.json and plugin manifest versions must match",
  );
  requireCondition(
    packageJson.name === `@kritama/${name}`,
    "npm package name must match the plugin name under the @kritama scope",
  );
  requireCondition(packageJson.license === "Apache-2.0", "package.json license must be Apache-2.0");
  requireCondition(
    !("mcpServers" in manifest) && !existsSync(join(root, ".mcp.json")),
    "Platform submission configures the Template MCP URL instead of bundling MCP configuration",
  );

  const pluginInterface = manifest.interface;
  requireCondition(isObject(pluginInterface), "interface must contain a JSON object");
  requireText(pluginInterface.displayName, "interface.displayName", 30);
  requireText(pluginInterface.shortDescription, "interface.shortDescription", 30);
  requireText(pluginInterface.longDescription, "interface.longDescription", 4000, {
    oneLine: false,
  });
  requireText(pluginInterface.developerName, "interface.developerName", 80);
  requireCondition(CATEGORIES.has(pluginInterface.category), "interface.category is unsupported");

  const capabilities = pluginInterface.capabilities ?? [];
  requireCondition(
    Array.isArray(capabilities) && capabilities.length <= 20,
    "interface.capabilities must contain at most 20 entries",
  );
  capabilities.forEach((capability, index) => {
    requireText(capability, `interface.capabilities[${index}]`, 120);
  });

  const prompts = pluginInterface.defaultPrompt ?? [];
  requireCondition(
    Array.isArray(prompts) && prompts.length <= 3,
    "interface.defaultPrompt must contain at most three entries",
  );
  const normalizedPrompts = new Set();
  prompts.forEach((prompt, index) => {
    const promptText = requireText(prompt, `interface.defaultPrompt[${index}]`, 128);
    requireCondition(
      !promptText.includes("@"),
      `interface.defaultPrompt[${index}] must not contain an app @mention`,
    );
    const normalized = promptText.trim().split(/\s+/u).join(" ").toLowerCase();
    requireCondition(
      !normalizedPrompts.has(normalized),
      "interface.defaultPrompt entries must be unique",
    );
    normalizedPrompts.add(normalized);
  });

  for (const field of ["websiteURL", "privacyPolicyURL", "termsOfServiceURL", "supportURL"]) {
    if (field in pluginInterface) {
      requireHttpsUrl(pluginInterface[field], `interface.${field}`);
    }
  }
  validateAsset(root, pluginInterface.logo, "interface.logo");
  validateAsset(root, pluginInterface.composerIcon, "interface.composerIcon");
  return [manifest, packageJson];
}

function validateCaseText(testCase, field, caseId) {
  requireText(testCase[field], `${caseId}.${field}`, 4000, { oneLine: false });
}

function validateMarketplace(root, manifest, packageJson) {
  const marketplace = readJson(join(root, ".agents", "plugins", "marketplace.json"));
  requireCondition(isObject(marketplace), "marketplace.json must contain a JSON object");
  requireCondition(Array.isArray(marketplace.plugins), "marketplace plugins must be a list");
  const matching = marketplace.plugins.filter(
    (entry) => isObject(entry) && entry.name === manifest.name,
  );
  requireCondition(matching.length === 1, "marketplace must contain exactly one plugin-name match");
  const source = matching[0].source;
  requireCondition(isObject(source), "marketplace source must be an object");
  requireCondition(source.source === "npm", "marketplace source must use npm");
  requireCondition(
    source.package === packageJson.name,
    "marketplace npm package must match package.json",
  );
  requireCondition(source.registry === "https://registry.npmjs.org", "npm registry is unsupported");
  const policy = matching[0].policy;
  requireCondition(isObject(policy), "marketplace policy must be an object");
  requireCondition(policy.installation === "AVAILABLE", "plugin must be available to install");
  requireCondition(
    policy.authentication === "ON_INSTALL",
    "plugin authentication must run on install",
  );
  requireCondition(matching[0].category === "Developer Tools", "marketplace category must match");
}

function validateSkillLayout(root) {
  const skillsRoot = join(root, "skills");
  const names = readdirSync(skillsRoot, { withFileTypes: true })
    .filter((entry) => entry.isDirectory() && existsSync(join(skillsRoot, entry.name, "SKILL.md")))
    .map((entry) => entry.name);
  requireCondition(
    names.length === 2 && names.includes("graph-builder") && names.includes("graph-audit"),
    "packaged skills must be exactly graph-builder and graph-audit",
  );
  for (const name of names) {
    const skillText = readFileSync(join(skillsRoot, name, "SKILL.md"), "utf8");
    requireCondition(
      new RegExp(`^name:\\s*${name}\\s*$`, "mu").test(skillText),
      `skills/${name}/SKILL.md name must match its folder`,
    );
    const agentText = readFileSync(join(skillsRoot, name, "agents", "openai.yaml"), "utf8");
    requireCondition(
      agentText.includes(`$${name}`),
      `skills/${name}/agents/openai.yaml must invoke $${name}`,
    );
  }
  return new Set(names);
}

function validateTemplateMcp(root, portal, manifest, reviewReady) {
  requireCondition(
    portal.submissionType === "MCP and Skills",
    "submissionType must be MCP and Skills",
  );
  const mcpServer = portal.mcpServer;
  requireCondition(isObject(mcpServer), "submission.mcpServer must be an object");
  requireCondition(mcpServer.urlMode === "Template", "mcpServer.urlMode must be Template");
  const template = requireText(
    mcpServer.templateMcpServerURL,
    "mcpServer.templateMcpServerURL",
    1024,
  );
  requireHttpsUrl(template, "mcpServer.templateMcpServerURL");
  const placeholders = placeholderNames(template);
  requireCondition(
    placeholders.length > 0,
    "Template MCP URL must contain at least one {name} placeholder",
  );
  requireCondition(
    placeholders.length === new Set(placeholders).size,
    "Template MCP URL placeholders must be unique",
  );
  requireCondition(
    template.replace(/\{[A-Za-z][A-Za-z0-9_]*\}/gu, "workspace") === "https://workspace/mcp",
    "Tama Kit Template MCP URL must keep the deployment host variable and /mcp path",
  );
  requireCondition(
    mcpServer.authentication === "OAuth 2.1",
    "mcpServer.authentication must be OAuth 2.1",
  );
  requireCondition(
    !("registeredAppId" in mcpServer),
    "mcpServer must not contain an integration or registered app ID",
  );
  requireCondition(
    !("apps" in manifest),
    "plugin.json apps is for local registered connections, not Platform submission",
  );
  requireCondition(
    !existsSync(join(root, ".app.json")),
    ".app.json is for local registered connections, not Platform submission",
  );

  const exampleUrl = mcpServer.exampleMcpServerURL;
  if (reviewReady) {
    requireCondition(
      exampleUrl !== null && exampleUrl !== undefined,
      "Example MCP Server URL is not configured; run npm run configure:mcp",
    );
  }
  if (exampleUrl === null || exampleUrl === undefined) {
    return;
  }
  const exampleText = requireText(exampleUrl, "mcpServer.exampleMcpServerURL", 1024);
  try {
    validatePublicExampleUrl(exampleText);
    validateTemplateMatch(template, exampleText);
  } catch (error) {
    throw new SubmissionError(error.message);
  }
}

function validateEvals(root, manifest, skillNames, reviewReady) {
  const data = readJson(join(root, "evals", "cases.json"));
  requireCondition(isObject(data), "evals/cases.json must contain a JSON object");
  const positive = data.positive_cases;
  const negative = data.negative_cases;
  requireCondition(
    Array.isArray(positive) && positive.length >= 5,
    "At least five positive test cases are required",
  );
  requireCondition(
    Array.isArray(negative) && negative.length >= 3,
    "At least three negative test cases are required",
  );

  const seenIds = new Set();
  for (const testCase of positive) {
    requireCondition(isObject(testCase), "Each positive test case must be an object");
    const caseId = requireText(testCase.id, "positive case id", 120);
    requireCondition(!seenIds.has(caseId), `Duplicate test case id: ${caseId}`);
    seenIds.add(caseId);
    validateCaseText(testCase, "prompt", caseId);
    const skillName = requireText(testCase.skill, `${caseId}.skill`, 120);
    requireCondition(skillNames.has(skillName), `${caseId}.skill must name a packaged skill`);
    requireCondition(
      Array.isArray(testCase.expected_behavior) && testCase.expected_behavior.length > 0,
      `${caseId}.expected_behavior must be a non-empty list`,
    );
    testCase.expected_behavior.forEach((item, index) => {
      requireText(item, `${caseId}.expected_behavior[${index}]`, 1000, { oneLine: false });
    });
    validateCaseText(testCase, "expected_result_shape", caseId);
    validateCaseText(testCase, "fixture_data", caseId);
  }

  for (const testCase of negative) {
    requireCondition(isObject(testCase), "Each negative test case must be an object");
    const caseId = requireText(testCase.id, "negative case id", 120);
    requireCondition(!seenIds.has(caseId), `Duplicate test case id: ${caseId}`);
    seenIds.add(caseId);
    validateCaseText(testCase, "prompt", caseId);
    validateCaseText(testCase, "expected_behavior", caseId);
    validateCaseText(testCase, "reason", caseId);
  }

  const portal = readJson(join(root, "submission", "portal.json"));
  requireCondition(isObject(portal), "submission/portal.json must contain a JSON object");
  validateTemplateMcp(root, portal, manifest, reviewReady);
  requireCondition(isObject(portal.listing), "submission listing must be an object");
  for (const field of ["websiteURL", "privacyPolicyURL", "termsOfServiceURL", "supportURL"]) {
    requireHttpsUrl(portal.listing[field], `submission.listing.${field}`);
  }
  validateCaseText(portal, "releaseNotes", "submission");

  const positiveIds = new Set(positive.map((testCase) => testCase.id));
  const negativeIds = new Set(negative.map((testCase) => testCase.id));
  const selectedPositive = portal.positiveCaseIds;
  const selectedNegative = portal.negativeCaseIds;
  requireCondition(
    Array.isArray(selectedPositive) && selectedPositive.length >= 5,
    "submission.positiveCaseIds must select at least five positive cases",
  );
  requireCondition(
    Array.isArray(selectedNegative) && selectedNegative.length >= 3,
    "submission.negativeCaseIds must select at least three negative cases",
  );
  selectedPositive.forEach((caseId, index) => {
    requireText(caseId, `submission.positiveCaseIds[${index}]`, 120);
  });
  selectedNegative.forEach((caseId, index) => {
    requireText(caseId, `submission.negativeCaseIds[${index}]`, 120);
  });
  requireCondition(
    selectedPositive.length === new Set(selectedPositive).size,
    "submission.positiveCaseIds must be unique",
  );
  requireCondition(
    selectedNegative.length === new Set(selectedNegative).size,
    "submission.negativeCaseIds must be unique",
  );
  requireCondition(
    selectedPositive.every((caseId) => positiveIds.has(caseId)),
    "submission.positiveCaseIds contains an unknown case",
  );
  requireCondition(
    selectedNegative.every((caseId) => negativeIds.has(caseId)),
    "submission.negativeCaseIds contains an unknown case",
  );
  return [positive.length, negative.length];
}

export function validateRepository(root, { reviewReady = false } = {}) {
  const resolvedRoot = resolve(root);
  const [manifest, packageJson] = validateManifest(resolvedRoot);
  validateMarketplace(resolvedRoot, manifest, packageJson);
  const skillNames = validateSkillLayout(resolvedRoot);
  const [positiveCount, negativeCount] = validateEvals(
    resolvedRoot,
    manifest,
    skillNames,
    reviewReady,
  );
  console.log(
    `Submission validation passed: ${manifest.name} ${manifest.version} ` +
      `(${positiveCount} positive, ${negativeCount} negative cases, ` +
      `review_ready=${String(reviewReady)})`,
  );
  return [manifest, packageJson];
}

function usage() {
  return "Usage: validate-submission.mjs [--review-ready] [root]";
}

export function main(argv = process.argv.slice(2)) {
  const { values, positionals } = parseArgs({
    args: argv,
    options: {
      "review-ready": { type: "boolean", default: false },
      help: { type: "boolean", short: "h" },
    },
    allowPositionals: true,
    strict: true,
  });
  if (values.help) {
    console.log(usage());
    return 0;
  }
  requireCondition(positionals.length <= 1, usage());
  const root = positionals[0] ?? ROOT_DEFAULT;
  validateRepository(root, { reviewReady: values["review-ready"] });
  return 0;
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  try {
    process.exitCode = main();
  } catch (error) {
    console.error(`Submission validation failed: ${error.message}`);
    process.exitCode = 1;
  }
}
